### PR TITLE
Improve property access performance

### LINKED ISSUE #50: Increase initialization timeout
We've starting seeing this exception error happen on and off, let's increase the time out 👍 
```
2021-05-24T22:48:48.7695468Z ERROR:: WorkerThread.<.ctor>b__7_0(): WorkerThread(): exception thrown when running task System.TimeoutException: Timeout while waiting for assemblies to load
   at Python.Runtime.AssemblyManager.Initialize()
   at Python.Runtime.Runtime.Initialize(Boolean initSigs, ShutdownMode mode)
   at Python.Runtime.PythonEngine.Initialize(IEnumerable`1 args, Boolean setSysArgv, Boolean initSigs, ShutdownMode mode)
   at Python.Runtime.PythonEngine.Initialize(Boolean setSysArgv, Boolean initSigs, ShutdownMode mode)
   at Python.Runtime.PythonEngine.Initialize()
   at QuantConnect.Python.PythonInitializer.Initialize() in /LeanCloud/CI.Builder/bin/Debug/src/QuantConnect/Lean/Common/Python/PythonInitializer.cs:line 42
   at QuantConnect.Util.WorkerThread.<.ctor>b__7_0() in /LeanCloud/CI.Builder/bin/Debug/src/QuantConnect/Lean/Common/Util/WorkerThread.cs:line 62
```

### PR DESCRIPTION
**Pull request by jmerle**
### What does this implement/fix? Explain your changes.

This PR improves property access performance considerably:
- Replaced GetValue/SetValue calls for property/field access with [FastMember](https://www.nuget.org/packages/FastMember)'s TypeAccessor when possible. This is considerably faster than reflection, but not supported in all cases. In cases where it's not supported we fall back to normal reflection.
- Replaced a ConcurrentDictionary in `PyObjectConversions` with a normal dictionary and added a `Count == 0` check before calling `TryGetValue` on it.
- Wrapped code related to tracking managed objects in an if so it only gets called when it's actually useful.

This PR adds 36 tests covering Python code getting/setting C# properties/fields. The changes pass these tests and all of LEAN's regression tests.

Unrelated changes:
- Doubled the assembly loading timeout to fix #50.
- Bumped the version to 2.0.2.

[IndicatorRibbonBenchmark](https://github.com/QuantConnect/Lean/blob/master/Algorithm.Python/Benchmarks/IndicatorRibbonBenchmark.py) benchmarks (we currently use a Python.NET debug build in production):
```
LEAN debug master // Python.NET debug master:
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 149,20 seconds at 11k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 146,29 seconds at 11k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 146,48 seconds at 11k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 148,64 seconds at 11k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 147,71 seconds at 11k data points per second. Processing total of 1.578.463 data points.
Average time: 147,664 seconds

LEAN debug master // Python.NET debug PR:
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 115,87 seconds at 14k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 115,61 seconds at 14k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 120,06 seconds at 13k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 119,92 seconds at 13k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 118,67 seconds at 13k data points per second. Processing total of 1.578.463 data points.
Average time: 118,026 seconds (-20.1% against Python.NET debug master)

LEAN debug master // Python.NET release master:
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 89,52 seconds at 18k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 88,79 seconds at 18k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 88,47 seconds at 18k data points per second. Processing total of 1.578.463 data points.
Average time: 88,93 seconds (-39.8% against Python.NET debug master)

LEAN debug master // Python.NET release PR:
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 65,42 seconds at 24k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 66,76 seconds at 24k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 66,83 seconds at 24k data points per second. Processing total of 1.578.463 data points.
Average time: 66,34 seconds (-55.1% against Python.NET debug master, -25.4% against Python.NET release master)
```

### Does this close any currently open issues?

Closes #50.

### Any other comments?

No.

### Checklist

Check all those that are applicable and complete.

-   [x] Make sure to include one or more tests for your change
-   [ ] If an enhancement PR, please create docs and at best an example
-   [ ] Add yourself to [`AUTHORS`](../blob/master/AUTHORS.md)
-   [ ] Updated the [`CHANGELOG`](../blob/master/CHANGELOG.md)

### CODE REVIEW

File: src/runtime/fieldobject.cs [hunk 1]
```diff
@@ -67,7 +69,18 @@ public static IntPtr tp_descr_get(IntPtr ds, IntPtr ob, IntPtr tp)
                     Exceptions.SetError(Exceptions.TypeError, "instance is not a clr object");
                     return IntPtr.Zero;
                 }
-                result = info.GetValue(co.inst);
+
+                // FastMember's TypeAccessor increases performance, but only works on public fields
+                var typeAccessor = info.IsPublic ? self.GetTypeAccessor(co.inst) : null;
+                if (typeAccessor != null)
```
Comment (Martin-Molinero) [inline src/runtime/fieldobject.cs:75]: `info.IsPublic` could directly be the condition of the if? same bellow

File: src/runtime/propertyobject.cs [hunk 2]
```diff
@@ -146,7 +158,17 @@ public new static int tp_descr_set(IntPtr ds, IntPtr ob, IntPtr val)
                         Exceptions.RaiseTypeError("invalid target");
                         return -1;
                     }
-                    info.SetValue(co.inst, newval, null);
+
+                    // FastMember's TypeAccessor increases performance, but only works on public properties
+                    var typeAccessor = setter.IsPublic ? self.GetTypeAccessor(co.inst) : null;
+                    if (typeAccessor != null)
```
Comment (Martin-Molinero) [inline src/runtime/propertyobject.cs:164]: same comment 👍

File: src/runtime/typeaccessormanager.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using FastMember;
+
+namespace Python.Runtime
+{
+    /// <summary>
+    /// TypeAccessorManager creates TypeAccessor instances when necessary.
+    /// </summary>
+    public class TypeAccessorManager
+    {
+        private static readonly Dictionary<Type, TypeAccessor> Cache = new Dictionary<Type, TypeAccessor>(128);
```
Comment (Martin-Molinero) [inline src/runtime/typeaccessormanager.cs:14]: nit - `readonly` adds a small performance overhead, let's remove it 👍

File: src/runtime/fieldobject.cs (hunk 1 above)
Comment (jmerle) [reply src/runtime/fieldobject.cs:75]: `self.GetTypeAccessor` can return `null`, that's why it's outside the if.

File: src/runtime/propertyobject.cs (hunk 2 above)
Comment (jmerle) [reply src/runtime/propertyobject.cs:164]: `self.GetTypeAccessor` can return `null`, that's why it's outside the if.

**Comment by jmerle** [comment]
Summary of the latest changes:
- Replaced FastMember with Fasterflect because it's slightly faster and because LEAN also uses Fasterflect.
- Re-added an `if (inst is Exception)` that was lost in a previous rebase (https://github.com/QuantConnect/pythonnet/blob/1.0.5.30/src/runtime/clrobject.cs#L33).

New benchmarks:
```
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 62,48 seconds at 25k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 62,13 seconds at 25k data points per second. Processing total of 1.578.463 data points.
Debug: Algorithm Id:(IndicatorRibbonBenchmark) completed in 63,80 seconds at 25k data points per second. Processing total of 1.578.463 data points.
Average time: 62,80 seconds (-57.8% against Python.NET debug master, -29.4% against Python.NET release master)
```

**Pull request merged**